A word-based binary integer encoding approach.

oscbyspro opened the pull request:
Salutations! I come in peace ✌️

> [!NOTE]
> This pull request builds on [an earlier contribution](https://github.com/apple/swift-foundation/pull/262).

### Description

These changes aim to achieve two things. The first is making the binary integer text encoding algorithm non-generic so that models such as DoubleWidth perform well. The second is performing mutating division, making arbitrary precision faster. Copying the binary integer's words to a temporary allocation achieves both. Since the new algorithm takes a collection of words, it can also test big integers without big integer models.

### Measurements

I have published a small [package][PRP] with some [benchmarks][PRT]. 

- `v0`: the `old` version
- `v1`: the `new` version
- `v2`: the `new` version + 2x `@inlinable` + 1x `@usableFromInline`

#### The results on my machine in seconds:

|         | v0     | v1    | v2    | [Numberick][NBK] |
|---------|-------:|------:|------:|-----------------:|
| Int     |  1.755 | 0.314 | 0.188 | 0.172            |
| UInt    |  0.261 | 0.302 | 0.181 | 0.166            |
| Int256  | 42.807 | 0.875 | 0.443 | 0.430            |
| UInt256 | 34.844 | 0.815 | 0.428 | 0.426            |

[PRP]: https://github.com/oscbyspro/my-swift-foundation-pull-request
[PRT]: https://github.com/oscbyspro/my-swift-foundation-pull-request/blob/main/Tests/PullRequestTests/Tests.swift
[NBK]: https://github.com/oscbyspro/Numberick

itingliu commented:
@wadetregaskis Do you want to take a look?

wadetregaskis commented:
> @wadetregaskis Do you want to take a look?

Yeah, I'm happy to, but it probably won't be until later this week.  I already looked over it a bit when it was sent to me originally, and I think overall it looks good.  I might have some questions, but don't wait on me, if everyone else is happy with this patch already.

iCharlesHu commented:
@swift-ci please test

wadetregaskis commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+        let remainder: UInt
+        (magnitude, remainder) = magnitude.quotientAndRemainder(dividingBy: 10)
+        buffer.base[index] = UInt8(ascii: "0") &+ UInt8(truncatingIfNeeded: remainder)
+    }   while magnitude != 0 as UInt
```
Is the `as UInt` necessary?  `magnitude` is defined as `Uint` (since it's initialised from `word` which is `UInt`), so there should be no ambiguity nor any significant type inference cost.

oscbyspro replied:
Dunno. Swift's BinaryInteger protocol comes with has generic comparison operators, Swift's preferred integer literal is Int, and Xcode does not show what integer literals resolve to.

oscbyspro replied:
I see that  puts a lot of emphesis on avoiding type assumptions. My concern was more about the expression resolving to an Int-to-UInt conversion. In any case, I've changed it to `.zero` since Xcode shows it resolving to the correct type.

wadetregaskis commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    var index = buffer.endIndex
+    
+    repeat {
+        precondition(index > buffer.startIndex)
```
`precondition` doesn't really log anything useful when it fails - it'd be wise to include a human-readable failure message with relevant context included.

wadetregaskis commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
````diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+/// 64-bit: (exponent: 19, power: 10000000000000000000)
+/// ```
+///
+private func maxDecimalExponentAndPowerForUnsignedIntegerWord() -> (exponent: Int, power: UInt) {
````
As a sidenote, this gets evaluated at compile time and reduced down to the actual constants, which is pretty awesome.  Its use later becomes two inline instructions, to load the immediates 0x8ac7230489e80000 and 0x13 (on x86-64).

wadetregaskis replied:
Yeah, I assume it's fine to rely on this compiler optimisation.  Plus I spent quite a while trying to come up with _any_ alternative that's remotely scalable and elegant, but found nothing.

That's also why in my earlier implementation I specifically commented about that, in the code.  I just hadn't discovered, at that point, that the compiler _does_ in fact optimise it, so I believe my comment was along the lines of "I _hope_ the compiler can optimise this…". 😁

wadetregaskis commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-            }
-
-            return indexOfHighestSetBitSeenSoFar + 1
+    let radix: (exponent: Int, power: UInt) = maxDecimalExponentAndPowerForUnsignedIntegerWord()
```
It might be better to fetch that in the closure below, as it appears the compiler doesn't automatically move it there, so this is incurring unnecessary closure argument passing overhead.

(which is minor - some wasted stack space and register juggling overhead, mainly - but might be precluding further optimisations through inlining)

wadetregaskis commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    // - Int.init(some BinaryFloatingPoint) rounds to zero.
+    // - Double.init(exactly:) and UInt.init(_:) for correctness.
+    // - log10(2.0) is: 1.0011010001000001001101⌈01...⌉ * 2^(-2).
+    return Int(Double(exactly: UInt(bitWidth))! * log10(2.0)) + 1
```
This doesn't work correctly for `bitWidth`s of 1 or less.  Those aren't values that will appear at runtime, but it might be worth putting an assert here to help ensure that just in case some future refactor exposes this to wider use.

And/or document this limitation in the function documentation.

Also, changing the parameter type to `UInt` will at least preclude negative values.

wadetregaskis replied:
I apologise, you're quite right.  I misunderstood the algorithm here (I think re. the rounding behaviour).  I've now tested it empirically and indeed it works as intended.

The value of 1 for bit width 0 is okay - that case shouldn't ever arise in practice anyway - although a `precondition(0 < bitWidth, …)` might be wise (or an assert).

oscbyspro replied:
The zero bit width case occurs when an integer's `words` collection is empty. I don't have an integer like that, but you could perhaps justify an unsigned big integer normalizing zero to `[]`. In any case, I have unit tests for empty collections, which result in `"0"`.

wadetregaskis reviewed (commented):
Looks pretty good overall, to me.

Since we're now focusing on performance, I spent a fair bit of time analysing the actual machine code (for x86-64) that this results in (in 'release' build configuration).  It's impressive what the compiler is able to do to reduce all this down to some pretty simple and short loops.

Although if I see it use `lea` to decrement a value by one again, I shall be very annoyed. 😜

wadetregaskis commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    var remainder = UInt.zero
+    
+    for index in (dividend.startIndex ..< dividend.endIndex).reversed() {
+        (dividend.base[index], remainder) = divisor.dividingFullWidth((high: remainder, low: dividend.base[index]))
```
It’s a shame that this results in the use of `udivti3` from compiler-rt instead of a simple `divq` instruction (on x86-64; presumably there’s an equivalent instruction on AArch64).  Evidently the compiler doesn’t believe that the result will never overflow (`udivti3` does the extra work of checking for and trapping on overflow, which is a waste here).

Perhaps there's some way to convince the compiler that the checks are unnecessary?  Overflow should be impossible in practice, by the inherent constraints of its use here.

oscbyspro replied:
That's an astute observation! I'd love to know if there's a better way to do it. I have an open issue in Numberick ([#59](https://github.com/oscbyspro/Numberick/issues/59)) about division being slower than expected in some cases. I wrote a note about udivmodti4 taking up most of the time in Instruments, but I did not know how to proceed beyond that. It's a bit funky that it performs oveflow checking when it truncates the result anyways:

```swift
print(UInt(123).dividingFullWidth((high: 122, low: UInt.max))) // (quotient: UInt.max, remainder: 122)
print(UInt(123).dividingFullWidth((high: 123, low: 00000000))) // (quotient: 00000000, remainder: 000)
```

wadetregaskis replied:
I was mistaken about `__udivti3` re. the overflow case.  It's doesn't actually trap, it just truncates, as you observed.  That's what one gets for trusting StackOverflow. 😝

For completeness… it's essentially using [`__udivmodti4`](https://github.com/apple/llvm-project/blob/76a94e518e019fa3766b8a95d61f46cc474695d1/compiler-rt/lib/builtins/udivmodti4.c#L100C24-L100C36) because `__udivti3` is basically an alias; it's two instructions: one to zero out r8 to effect a `NULL` for the out parameter that reports the remainder, and then a jump to `__udivmodti4` (which incidentally is a waste of space since `__udivmodti4` is the _very_ next procedure, so it'd flow into it automatically without the `jmp`).

`__udivmodti4` is a little complicated, both at the C level and in the resulting assembly (67 instructions, on x86-64, and branchy).  It does use `divq`, ultimately (it's actually hard-coded to on x86-64, via inline assembly).  But you still have the overhead in the function call and unnecessary checks to use other paths - it really would benefit from being inlined.  I don't know how to make the compiler inline it, though.

Which raises the question:  why on earth isn't it just using `__udivmodti4` directly, since we _do_ care about the remainder here.  It appears that it's using the result of the division to calculate the remainder manually:

```assembly
cmp  r12, rdi ; Check that the index is > 0.  On first entry to this loop it's endIndex.
jle  loc_19eb ; Basically, crash if index is <= 0.
lea  rbx, qword [r12-1] ; Decrement index of current word in `words`.
mov  r13, qword [r15+r12*8-8] ; Load next word from `words` argument.
mov  rdi, r13 ; low half of 128-bit argument "a" (dividend) for __udivti3.  rsi is high half, and is initially set to zero but is the remainder in subsequent iterations.
mov  rdx, r14 ; low half of 128-bit argument "b" (divisor) for __udivti3.
xor  ecx, ecx ; high half of argument "b".  Top 32 bits are known zero from earlier in the function.
call __udivti3 ; rax = a / b
mov  rdi, qword [rbp+var_38] ; words.count.
mov  rcx, rax
mul  r14 ; Quotient * divisor (the function argument, not the argument to __udivti3.
mov  rsi, r13 ; The aforementioned next word from `words`.
sub  rsi, rax ; Remainder = current word - (Quotient * divisor)
mov  qword [r15+r12*8-8], rcx ; Save the quotient over the current word in `words`.
mov  r12, rbx ; Just shifting the current index around registers.
cmp  rdi, rbx ; Check that index is != words.count.
jne  loc_19a0 ; Repeat this whole block while it isn't.
```

I don't see how that could be faster than using `__udivmodti4`, since the latter calculates the remainder no matter what - it merely drops it silently if its 'rem' out argument is `NULL`, as it is in the use via `__udivti3`.  Weird.  Seems like an optimiser bug.

oscbyspro commented:
Oh, neat. I appreciate your feedback. I'll address each point in time.

>[!NOTE]
> The algorithm is fundamentally sign-magnitude, and it would be simpler to call `magnitude.words` and skip the form-two's-complement method. I have tried it, and it performs similarly when inlinable, but DoubleWidth\<T\> performs 3x worse when warring against module boundaries compared calling `self.words`.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+///
+/// - Parameter words: A binary integer's mutable words.
+///
+private func formTwosComplementForBinaryInteger(words: UnsafeMutableBufferPointer<UInt>) {
```
I don't have a strong opinion about it, but I found your comment a bit perplexing because I don't ascribe signedness to it. The 2's complement format is named after the 2's complement operation `~x &+ 1` (unsigned, fixed-width, `2 ^ x.bitWidth - x`). It took me a while to pinpoint, but I believe this snippets highlight our different viewpoints:


```swift
for x in Int8.min ... Int8.max {
    var magnitude = UInt8(bitPattern: x) // some bit pattern
    
    if type(of: x).isSigned && magnitude >> (x.bitWidth - 1) == 1 {
        magnitude = ~magnitude &+ 1 // form 2's complement when needed
    }
    
    precondition(x.magnitude == magnitude) // bit pattern is magnitude
}

// MARK: Dunno, but how I think you see it:

for x in Int8.min ... Int8.max {
    let magnitude: UInt8
    
    magnitude = UInt8(bitPattern: x < 0 ? ~x &+ 1 : x) // some representation conversion
    
    precondition(x.magnitude == magnitude)
}
```

wadetregaskis replied:
I think the key point is that you can't fit `x.magnitude` into `Self` for any signed `Self`, specifically because of `type(of: x).min`.  e.g. for `Int8` (just to keep the numbers small) `Int8.min` is -128.  The inverse of that is 128, which is greater than `Int8.max` (of 127).  You have to upgrade to `Int16` to represent it _in two's complement form_.

Aside: this is presumably why the type of the `magnitude` property is _not_ `Self`, but rather an associated value defined by each specific type.  All the built-in signed integer types use their unsigned partners for it.

So, the function in question _cannot_ perform two's complement correctly for all values; it's not a generic two's complement implementation.  What it _can_ do is:

1. Invert any _positive_ two's complement-encoded value.
2. Invert any negative, two's complement-encoded value into its positive form _as an unsigned-type of the same bit width_.

(1) isn't relevant to the function's actual use in this file, only (2).  Thus why I suggested a name that is more specific to its intended use (even if it technically ignores another use it can work for, (1)). 

This is all just to try to clarify.  I don't feel that this is a particular important point in the grand scheme of things, either.  It just seems like a sensible little refinement, to me.

wadetregaskis replied:
Or at least, not for _all_ signed `BinaryInteger`s.  You can certainly have some where that limitation doesn't apply - e.g. because they're flexible-width - or I suppose hypothetically even a `FixedWidthInteger` that constraints it minimum value to something less (in magnitude) than 2^(bitWidth-1).  But the code in question has to work for _all_ `BinaryInteger`s.

oscbyspro replied:
It performs the *two's complement operation* `~x &+ 1`, which is defined for all binary integers. It's signedness agnostic, and it does not reinterpret anything. Whether the result of `~x &+ 1` is useful to us is determined at the call site.

oscbyspro replied:
This is only to clarify that you cannot find an input where `~x &+ 1` does not fit. In the one place it's used, the `words` has—to your point—been reinterpreted as an unsigned `magnitude`. The type system isn't there to help, so there's no bit cast reinterpreting it. I get the feeling that you'd be happier with renaming the `magnitude` as `reinterpretedUnsignedMagnitude` or somesuch, instead of renaming this method :man_shrugging:

oscbyspro replied:
I changed the call site to the following for clarity:

```swift
//  We reinterpret the words as an unsigned binary integer.
var magnitude = /* consume */ words
//  Note that negative values are in two's complement form.
let isLessThanZero = isSigned && Int(bitPattern: magnitude.last ?? .zero) < .zero
//  The **unsigned** magnitude is formed when the words represent a negative value.
if  isLessThanZero {
    formTwosComplementForBinaryInteger002(words: magnitude)
}
```

itingliu commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+        asciiIndex = writeIndex
+        
+        //  Add a minus sign to negative values.
+        if  isLessThanZero {
```
```suggestion
        if isLessThanZero {
```

itingliu commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+        
+        //  Add a minus sign to negative values.
+        if  isLessThanZero {
+            ascii.formIndex(before:  &asciiIndex)
```
```suggestion
            ascii.formIndex(before: &asciiIndex)
```

itingliu commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+///   - words: The binary integer's words.
+///   - isSigned: The binary integer's signedness.
+///
+internal // In order to test big integers without big integer models.
```
```suggestion
internal
```

Nit: I'd prefer leave out this kind of comment, especially when it's already an established pattern throughout the project

itingliu commented:
@swift-ci please test

itingliu commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+///   - isSigned: The binary integer's signedness.
+///
+internal // In order to test big integers without big integer models.
+func numericStringRepresentationForBinaryInteger(
```
Would it make sense to declare this inside an extension on `BinaryInteger` so that we can have `words: Self.Words` instead of it being an existential `some Collection<UInt>`? Also a bonus point for less cluttering of global namespace?

oscbyspro replied:
You would lose the ability to unit test it without adding or importing a big integer model. Big integers, as defined by Swift's BinaryInteger protocol, are basically fancy `some Collection<UInt>`. I added this method to allow unit testing using `Array<UInt>` since the need for external big integers was an earlier complaint. The other helper methods are single-use,  however. So you can merge them into one intimidating blob if clutter is an issue.

itingliu replied:
Ah, that makes sense. So, that was what your previous comment was about, but you removed because of my ask, sorry 😬

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    // Copies the words and then passes them to a non-generic, mutating, word-based algorithm.
+    withUnsafeTemporaryAllocation(of: UInt.self, capacity: words.count) { copy in
+        let count =  copy.initialize(fromContentsOf: words)
+        defer{ copy.baseAddress!.deinitialize(count: count) }
```
I changed it to the following since [withUnsafeTemporaryAllocation(of:capacity:)](https://github.com/apple/swift/blob/b72f0c95e4f711158b9b1c015cc22fa6653b2c86/stdlib/public/core/TemporaryAllocation.swift#L335) does not guarantee the size of the buffer (although the current implementation covers exactly the requested number of elements).

```swift
internal func numericStringRepresentationForBinaryInteger(
words: some Collection<UInt>, isSigned: Bool) -> ArraySlice<UInt8> {
    // Copies the words and then passes them to a non-generic, mutating, word-based algorithm.
    withUnsafeTemporaryAllocation(of: UInt.self, capacity: words.count) {
        let initializedEndIndex = $0.initialize(fromContentsOf: words)
        let initialized = UnsafeMutableBufferPointer(rebasing: $0[..<initializedEndIndex])
                
        defer {
            initialized.deinitialize()
        }
        
        return numericStringRepresentationForMutableBinaryInteger(words: initialized, isSigned: isSigned)
    }
}
```

oscbyspro commented:
Hm. The documentation of [Swift.withUnsafeTemporaryAllocation(of:capacity:)](https://developer.apple.com/documentation/swift/withunsafetemporaryallocation(of:capacity:_:)) states:

> The implementation may allocate a larger buffer pointer than is strictly necessary to contain capacity values of type type. The behavior of a program that attempts to access any such additional storage is undefined.

Does that speak of the buffer in the closure, or is it just implementation details leaking through documentation?

itingliu reviewed (commented):
@glessard Do you want to take a look, specifically about the usage around the `unsafemutableBufferPointer`? And also if we should consider sinking these into stdlib?

itingliu commented:
@swift-ci please test

wadetregaskis commented:
> Hm. The documentation of [Swift.withUnsafeTemporaryAllocation(of:capacity:)](https://developer.apple.com/documentation/swift/withunsafetemporaryallocation(of:capacity:_:)) states:
> 
> > The implementation may allocate a larger buffer pointer than is strictly necessary to contain capacity values of type. The behavior of a program that attempts to access any such additional storage is undefined.
> 
> Does that speak of the buffer in the closure, or is it just implementation details leaking through documentation?
> 
> It sounds like it refers to the buffer in the closure because that's how Swift uses the term **capacity** elsewhere:
> 
> ```swift
> for capacity in 0 ... 32 {
>     String(unsafeUninitializedCapacity: capacity) {
>         print("capacity: \(capacity), count: \($0.count)")
>         precondition($0.count == Swift.max(16, capacity))
>         return 0
>     }
> }
> ```

I believe it means that `$0.count` is guaranteed to be >= `capacity`, not merely ==.

It makes sense as such.  Hypothetically for a capacity <= 15 it can provide a buffer pointing to the actual stack space allocated for the `String` (using the `String`'s "inline" form to avoid a heap allocation), so it might as well just pass 15 in that case.  And of course for its non-inline forms its heap allocations have to be multiples of 16 at least (malloc on Apple platforms has never supported allocation granularities smaller than 16 bytes).

It turns out [the current implementation](https://github.com/apple/swift/blob/cabb5e109fa373ff82b3d335fd6055d93ada98d8/stdlib/public/core/String.swift#L560) _does_ do this optimisation - of directly initialising the stack memory for capacities <= 15 - but oddly it doesn't pass the _actual_ capacity in that case (which would be 15), instead pretending it's always exactly the requested capacity.  An oversight, perhaps?

Talking in 64-bit here.  32-bit the same principles apply but the numbers might be different.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+func numericStringRepresentationForBinaryInteger(
+words: some Collection<UInt>, isSigned: Bool) -> ArraySlice<UInt8> {
+    // Copies the words and then passes them to a non-generic, mutating, word-based algorithm.
+    withUnsafeTemporaryAllocation(of: UInt.self, capacity: words.count) { copy in
```
Hm. That is fair. I should measure some small values and perhaps retrofit the fast path you wrote. As a side note, it is possible to return a String instead of an ArraySlice through an additional temporary allocation. Given that String and withUnsafeTemporaryAllocation(\_:) can stack allocate, it might be beneficial. I haven't measured the difference, however.

oscbyspro commented:
I have added a signed and unsigned tiny integer benchmark to see whether fast pathing is worth it. I've also measured an `@inlinable` version of the original implementation to see what is and isn't a module boundary issue. The original fast path does, indeed, improve the non-`@inlinable` performance of tiny integers, but Swift's standard library beats it. It seems, to me, that returning a String and stack allocating small integers is more impactful.

> [!NOTE]
> I've changed to Xcode 15.0.1, so everything is faster.

glessard commented:
> > The implementation may allocate a larger buffer pointer than is strictly necessary to contain capacity values of type. The behavior of a program that attempts to access any such additional storage is undefined.
> 
> Does that speak of the buffer in the closure, or is it just implementation details leaking through documentation?

I think this is a leak of implementation details, coupled with an API documentation deserving some improvements.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,187 @@ extension BinaryInteger {
+        
+        // Add a minus sign to negative values.
+        if isLessThanZero {
+            ascii.formIndex(before: &asciiIndex)
```
Sorry, I didn't see your question earlier. The `capacity` is given by the following expression so please have a look at [number-of-decimal-digits-in-a-binary-integer](https://www.exploringbinary.com/number-of-decimal-digits-in-a-binary-integer). 

```swift
let capacity = maxDecimalDigitCountForUnsignedInteger(bitWidth: words.count * UInt.bitWidth) + (isLessThanZero ? 1 : 0)
```

Since `capacity` is only an upper bound, I've used the fullest possible width as an uncomplicated approximaton (it can be trimmed a bit). In the loop, we then divide by `10` repeatedly and write each digit back-to-front. The loop is on `repeat`, so we write at least one. The minimum capacity is `1` when `words.count == 0`, and `1` is added when `isLessThanZero == true`.

```
let isLessThanZero = isSigned && Int(bitPattern: words.last ?? 0) < 0 // tests the last bit when isSigned == true
```

#### Important edge case thing-y

The loop branchlessly decrements `asciiIndex` by `radix.exponent` (the digits written plus zeros for padding) each iteration, so it might be negative at the end of the loop. This is one reason why we set it to `writeIndex` afterwards, the other is trimmings the last iteration's padding. This assumes `Index == Int` (alternatives exist).

oscbyspro replied:
For what it's worth, no index goes out of bounds or assumes typing in v2. I've also replaced all assertions with preconditions since the difference in performance was insignificant.

oscbyspro commented:
Quick update: I'm working on a new version that should be speedy for small integer too. I'll update this pull request within the next few days or so. It looks promising. It involves returning a String instead of an ArraySlice, so I hope that's OK. The ASCII sequence can be accessed using `withUTF8(_:)`.

oscbyspro commented:
As a side note, I saw that BinaryInteger's `formatted()` method casts to Int, which traps big integers:

```swift
extension BinaryInteger {
    /// Format `self` using `IntegerFormatStyle()`
    public func formatted() -> String {
        IntegerFormatStyle().format(Int(self))
    }
}
```

oscbyspro commented:
I have added the 1,000,000th Fibonacci sequence element (210k decimal digits) to the table :smiley:

glessard commented:
@swift-ci please test

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    let isLessThanZero = isSigned && Int(bitPattern: words.last ?? 0) < 0
+    //  The magnitude is formed when the words represent a negative value.
+    if  isLessThanZero {
+        formTwosComplementForBinaryInteger(words: words)
```
While it's possible to move this transformation to the `BinaryInteger` extension, it's only viable if `@inlinable` because it won't be specialized otherwise. The unspecialized performance penalty of evaluating something like `Self.isSigned && self < Self.zero` is quite apparent for `Int256`, but it can also seen in `Int`. If you are all OK with `@inlinable`, removing this method and calling `magnitude.words` is the way to go about it.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+///   - buffer: A buffer with enough memory to accommodate the format.
+///
+private func formTrailingNumericStringRepresentationForUnsignedInteger(
+word: UInt, into buffer: Slice<UnsafeMutableBufferPointer<UInt8>>) -> Int {
```
I dropped this function and inlined its body in `v2` because it was a bit arcane.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
-            return indexOfHighestSetBitSeenSoFar + 1
+    let radix: (exponent: Int, power: UInt) = maxDecimalExponentAndPowerForUnsignedIntegerWord()
+    let capacity = maxDecimalDigitCountForUnsignedInteger(bitWidth: words.count * UInt.bitWidth) + (isLessThanZero ? 1 : 0)
+    var ascii = ContiguousArray(repeating: UInt8(ascii: "0"), count: capacity) // Set initial ASCII zeros (see later steps).
```
It's simpler, and it removes a branch from the loop. I tried it both ways and did not see a benefit to in-loop fiddling.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    var power = 0010 as UInt
+    
+    while true {
+        let next = power.multipliedReportingOverflow(by: 10)
```
...and doing nothing saves 10 characters.

oscbyspro commented on Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift:
```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+    let capacity = maxDecimalDigitCountForUnsignedInteger(bitWidth: words.count * UInt.bitWidth) + (isLessThanZero ? 1 : 0)
+    var ascii = ContiguousArray(repeating: UInt8(ascii: "0"), count: capacity) // Set initial ASCII zeros (see later steps).
+    
+    var wordsIndex = words.endIndex
```
Slicing sounds reasonable, even if it does not change the algorithm. I've made it so the magnitude is sliced (rebased) instead of indexed. It's a bit slower, but not by much, and I'm not sure why. It was an opportunity to reinterpret the words as an unsigned binary integer, at least by variable name, so it's probably worth it for clarity. The break condition is now `magnitude.isEmpty`, for example.

oscbyspro commented:
I sense no enthusiasm—and nobody has made it a priority—should I close it? I know  has a habit of keeping things in limbo, but I don't. I'd much rather cross this PR off my list and go somewhere my contributions are appreciated :man_shrugging:

itingliu reviewed (approved):
Apologies for the delayed reply. I personally do want to take this one ✌️ 

@swift-ci please test

itingliu commented:
@swift-ci please test
